src/graphics/Window.zig

```diff
@@ -349,17 +583,174 @@ pub fn init() void { // MARK: init()
 	c.glEnable(c.GL_DEBUG_OUTPUT_SYNCHRONOUS);
 	c.glDebugMessageCallback(GLFWCallbacks.glDebugOutput, null);
 	c.glDebugMessageControl(c.GL_DONT_CARE, c.GL_DONT_CARE, c.GL_DONT_CARE, 0, null, c.GL_TRUE);
+	if (!settings.askToDownloadControllerMappings) {
+		downloadControllerMappings();
+	}
+	updateControllerMappings();
+	gamepadState = std.AutoHashMap(c_int, *c.GLFWgamepadstate).init(main.globalAllocator.allocator);
+	updateGamepadState();
+	std.log.debug("Gamepads at init: {d}", .{gamepadState.?.count()});
+}
+pub fn isControllerConnected() bool {
+	return gamepadState.?.count() > 0;
+}
+var controllerMappingsDownloaded: bool = false;
+var downloadControllerMappingsThread: ?std.Thread = null;
+pub fn controllerMappingsDownloading() bool {
+	return !controllerMappingsDownloaded and downloadControllerMappingsThread != null;
+}
+fn downloadControllerMappingsThreadFunc(curTimestamp: i128) void {
+	var client: std.http.Client = .{.allocator = main.globalAllocator.allocator};
+	var list = std.ArrayList(u8).init(main.globalAllocator.allocator);
+	if (client.fetch(.{
+		.method = .GET,
+		.location = .{.url = "https://raw.githubusercontent.com/mdqinc/SDL_GameControllerDB/master/gamecontrollerdb.txt"},
+		.response_storage = .{ .dynamic = &list }
+	}) catch null) |_| {
+		if (files.openDir(".") catch null) |dir| {
+			dir.write("gamecontrollerdb.txt", list.items) catch unreachable;
```

Comment:
This error can be reached, if e.g. `gamecontrollerdb.txt` already exists but the user doesn't have write permissions for whatever reason. Please properly handle errors like this.

Reply:
Also you can use `main.files.write()` to avoid all this boilerplate